src/workerd/jsg/ser.c++

```diff
@@ -41,22 +51,73 @@ v8::Maybe<uint32_t> Serializer::GetSharedArrayBufferId(
   return v8::Just(n);
 }
 
+void Serializer::scheduleDataCloneErrorForObject(v8::Isolate* isolate, v8::Local<v8::Object> obj) {
+  jsg::Lock& js = jsg::Lock::from(isolate);
+
+  try {
+    // The default error that V8 would generate is "#<TypeName> could not be cloned." -- for some
+    // reason, it surrounds the type name in "#<>", which seems bizarre? Let's generate a better
+    // error.
+    v8::Local<v8::String> message = js.str(kj::str(
+        "Could not serialize object of type \"", obj->GetConstructorName(), "\". This type does "
+        "not support serialization."));
+    isolate->ThrowException(makeDOMException(isolate, message, "DataCloneError"));
+  } catch (JsExceptionThrown&) {
+    // Apparently an exception was thrown during the construction of the DOMException. Most likely
+    // we were terminated. In any case, we'll let that exception stay scheduled and propagate back
+    // to V8.
+  }
+}
+
 void Serializer::ThrowDataCloneError(v8::Local<v8::String> message) {
-  // makeDOMException could throw an exception. If it does, we'll end up crashing but that's ok?
   auto isolate = v8::Isolate::GetCurrent();
-  isolate->ThrowException(makeDOMException(isolate, message, "DataCloneError"));
+  try {
+    isolate->ThrowException(makeDOMException(isolate, message, "DataCloneError"));
+  } catch (JsExceptionThrown&) {
+    // Apparently an exception was thrown during the construction of the DOMException. Most likely
+    // we were terminated. In any case, we'll let that exception stay scheduled and propagate back
+    // to V8.
+  }
+}
+
+bool Serializer::HasCustomHostObject(v8::Isolate* isolate) {
+  // V8 will always call WriteHostObject() for objects that have internal fields. We only need
+  // to override IsHostObject() if we want to treat pure-JS objects differently, which we do if
+  // treatClassInstancesAsPlainObjects is false.
+  return !treatClassInstancesAsPlainObjects;
+}
+
+v8::Maybe<bool> Serializer::IsHostObject(v8::Isolate* isolate, v8::Local<v8::Object> object) {
+  // This is only called if HasCustomHostObject() returned true.
+  KJ_ASSERT(!treatClassInstancesAsPlainObjects);
+  KJ_ASSERT(!prototypeOfObject.IsEmpty());
+
+  // If the object's prototype is Object.prototype, then it is a plain object, which we'll allow
+  // to be serialized normally. Otherwise, it is a class instance, which we should treat as a host
+  // object. Inside `WriteHostObject()` we will throw DataCloneError due to the object not having
+  // internal fields.
+  return v8::Just(object->GetPrototype() != prototypeOfObject);
```

Comment:
non-blocking for later: including a test to ensure that objects with null prototypes are properly handled would likely be good.

Reply:
89d996ab